src/navigators/createNavigator.js

```diff
@@ -1,19 +1,49 @@
 import React from 'react';
 
-/**
- * Creates a navigator based on a router and a view that renders the screens.
- */
-export default function createNavigator(router, routeConfigs, navigatorConfig) {
-  return NavigationView => {
-    class Navigator extends React.Component {
-      static router = router;
-      static navigationOptions = null;
+import getChildEventSubscriber from '../getChildEventSubscriber';
+import addNavigationHelpers from '../addNavigationHelpers';
 
-      render() {
-        return <NavigationView {...this.props} router={router} />;
-      }
-    }
+function createNavigator(NavigatorView, router, navigationConfig) {
+  class Navigator extends React.Component {
+    static router = router;
+    static navigationOptions = null;
+
+    render() {
+      const { navigation, screenProps } = this.props;
+      const { dispatch, state, addListener } = navigation;
+      const { routes } = state;
+
+      const descriptors = {};
+      routes.forEach(route => {
+        const getComponent = () =>
+          router.getComponentForRouteName(route.routeName);
 
-    return Navigator;
-  };
+        const childNavigation = addNavigationHelpers({
+          dispatch,
+          state: route,
+          addListener: getChildEventSubscriber(addListener, route.key),
+        });
+        const options = router.getScreenOptions(childNavigation, screenProps);
+        descriptors[route.key] = {
+          key: route.key,
+          getComponent,
+          options,
+          state: route,
+          navigation: childNavigation,
+        };
+      });
+
+      return (
+        <NavigatorView
+          screenProps={screenProps}
+          navigation={navigation}
+          navigationConfig={navigationConfig}
```

Comment:
It'll be nice to accept the `navigationConfig` as props to the component too. Which overrides the config defined statically.

Reply:
this seems reasonable to me, but I'm curious what @ericvicenti thinks.

one thing that strikes me as a bit weird about passing props in to navigators is that there is not really a good way to do this for anything but the root navigator. the only thing I can think of is:

```js
@connect(getSomeReduxState)
class MyNavigatorWithProps extends React.Component {
  static router = OriginalNavigator.router;

  render() {
    return (
      <OriginalNavigator
          myProp={this.props.someThingFromReduxState}
       />
    );
  }
}
```

is this what you had in mind @satya164? or do you think this is mainly just useful on the root navigator?

Reply:
I can't think of immediate use cases for it, but I agree. It makes a lot of sense to let props override the navigationConfig here.

Reply:
one concern I have is that we need to be sure that the props actually *can* be changed dynamically. things like `initialRouteName` aren't a big deal because you never expect them to change after render... but it's not clear that it is sane for us to support changing most navigator config props dynamically.